File: ios/ScannerUtils.swift
import Foundation
import MLKitBarcodeScanning

struct Size {
    let width: Int
    let height: Int
}

struct Ratio {
    let width: CGFloat
    let height: CGFloat
}

struct BoxRatio {
    let leftRatio: CGFloat
    let topRatio: CGFloat
    let widthRatio: CGFloat
    let heightRatio: CGFloat

    init(_ leftRatio: CGFloat, _ topRatio: CGFloat, _ widthRatio: CGFloat, _ heightRatio: CGFloat) {
        self.leftRatio = leftRatio
        self.topRatio = topRatio
        self.widthRatio = widthRatio
        self.heightRatio = heightRatio
    }
}

struct Scaling {
    let scale: CGFloat
    let dx: CGFloat
    let dy: CGFloat

    func destructured() -> (CGFloat, CGFloat, CGFloat) {
        return (scale, dx, dy)
    }
}

struct BoundingBox {
    var width: CGFloat
    var height: CGFloat
    var left: CGFloat
    var top: CGFloat
    var right: CGFloat
    var bottom: CGFloat
}

class ScannerUtils {
    static func getOptionsRatio(options: [AnyHashable: Any]?) -> Ratio {
        guard
            let ratio = options?["ratio"] as? [String: Any]
        else {
            return Ratio(width: 1.0, height: 1.0)
        }

        let width = ratio["width"] as? CGFloat ?? 1.0
        let height = ratio["height"] as? CGFloat ?? 1.0
        return Ratio(
            width: min(max(width, 0.0), 1.0),
            height: min(max(height, 0.0), 1.0)
        )
    }

    static func getOptionsOrientation(options: [AnyHashable: Any]?) -> UIDeviceOrientation {
        guard
            let orientation = options?["orientation"] as? String
        else {
            return .portrait
        }

        switch orientation {
            case "portrait": return .portrait
            case "landscape-left": return .landscapeLeft
            case "portrait-upside-down": return .portraitUpsideDown
            case "landscape-right": return .landscapeRight
            default: return .portrait
        }
    }

    static func getOptionsViewSize(options: [AnyHashable: Any]?) -> Size? {
        guard let viewSize = options?["viewSize"] as? [String: Any],
              let width = viewSize["width"] as? NSNumber,
              let height = viewSize["height"] as? NSNumber else {
            return nil
        }
        return Size(width: width.intValue, height: height.intValue)
    }

    static func getOptionsBarcodeFormats(options: [AnyHashable: Any]?) -> [Any] {
        if let formats = options?["formats"] as? [Any], !formats.isEmpty {
            return formats
        } else {
            return ["all"]
        }
    }

    static func getSafeBarcodeFormats(formats: [Any]) -> [BarcodeFormat] {
        return formats.compactMap { format -> BarcodeFormat in
            if let format = format as? String {
                switch format {
                    case "code_128": return .code128
                    case "code_39": return .code39
                    case "code_93": return .code93
                    case "codabar": return .codaBar
                    case "ean_13": return .EAN13
                    case "ean_8": return .EAN8
                    case "itf": return .ITF
                    case "upc_e": return .UPCE
                    case "upc_a": return .UPCA
                    case "qr": return .qrCode
                    case "pdf_417": return .PDF417
                    case "aztec": return .aztec
                    case "data_matrix": return .dataMatrix
                    case "all": return .all
                    default: return .all
                }
            } else {
                return .all
            }
        }
    }

    // ViewSize is used for scaling in the code.
    // The UI side should always be assumed as portrait.
    static func getSafeViewSize(imageSize: Size, viewSize: Size?) -> Size {
        if let viewSize = viewSize {
            let width = min(viewSize.width, viewSize.height)
            let height = max(viewSize.width, viewSize.height)
            return Size(width: width, height: height)
        } else {
            return imageSize
        }
    }

    static func getImageRotation(imageRotation: UIImage.Orientation) -> UIImage.Orientation {
        switch imageRotation {
            case .up: return .up
            case .left: return .right
            case .down: return .down
            case .right: return .left
            default: return .up
        }
    }

    static func getImageSizeBasedOnRotation(imageSizeRaw: Size, imageRotation: UIImage.Orientation) -> Size {
        switch imageRotation {
            case .left, .leftMirrored, .right, .rightMirrored:
                return Size(width: imageSizeRaw.height, height: imageSizeRaw.width)
            default:
                return imageSizeRaw
        }
    }

    static private func getImageScalingBasedOnViewSize(imageSize: Size, viewSize: Size) -> Scaling {
        let imageWidth = CGFloat(imageSize.width)
        let imageHeight = CGFloat(imageSize.height)
        let imageAspect = imageWidth / imageHeight

        let viewWidth = CGFloat(viewSize.width)
        let viewHeight = CGFloat(viewSize.height)
        let viewAspect = viewWidth / viewHeight

        let scale: CGFloat
        let scaledImageWidth: CGFloat
        let scaledImageHeight: CGFloat
        let dx: CGFloat
        let dy: CGFloat
        if imageAspect > viewAspect {
            // Image is wider -> crop left/right
            scale = viewHeight / imageHeight
            scaledImageWidth = imageWidth * scale
            scaledImageHeight = viewHeight
            dx = (scaledImageWidth - viewWidth) / 2.0
            dy = 0.0
        } else if viewAspect > imageAspect {
            // Image is taller → crop top/bottom
            scale = viewWidth / imageWidth
            scaledImageWidth = viewWidth
            scaledImageHeight = imageHeight * scale
            dx = 0.0
            dy = (scaledImageHeight - viewHeight) / 2.0
        } else {
            // Full compatibility -> crop nothing
            // Default if user does not provide viewSize value
            scale = 1.0
            scaledImageWidth = viewWidth
            scaledImageHeight = viewHeight
            dx = 0.0
            dy = 0.0
        }

        return Scaling(scale: scale, dx: dx, dy: dy)
    }


    // This transformation is necessary.
    // Because the default image always comes as landscapeRight.
    // On the Swift side, MLKit does not produce coordinates for a different orientation.
    // This step is done automatically on the Kotlin side.
    static func getRectBasedOnRotation(
        rect: CGRect,
        imageSize: Size,
        imageRotation: UIImage.Orientation
    ) -> CGRect {
        // By default, the camera captures data in Landscape mode.
        // Here it should always be width > height.
        // imageSize to imageRawSize conversion.
        let imageWidth = CGFloat(max(imageSize.width, imageSize.height))
        let imageHeight = CGFloat(min(imageSize.width, imageSize.height))

        switch imageRotation {
        case .up:
            return rect
        case .left:
            return CGRect(
                x: rect.minY,
                y: imageWidth - rect.maxX,
                width: rect.height,
                height: rect.width
            )
        case .right:
            return CGRect(
                x: imageHeight - rect.maxY,
                y: rect.minX,
                width: rect.height,
                height: rect.width
            )
        case .down:
            return CGRect(
                x: imageWidth - rect.maxX,
                y: imageHeight - rect.maxY,
                width: rect.width,
                height: rect.height
            )
        default:
            return rect
        }
    }

    static func getBoundingBoxFromRect(rect: CGRect) -> BoundingBox {
        return BoundingBox(
            width: rect.width,
            height: rect.height,
            left: rect.minX,
            top: rect.minY,
            right: rect.maxX,
            bottom: rect.maxY
        )
    }

    static func getBoxRatioBasedOnOrientationForUI(
        boxScaled: BoundingBox,
        viewSize: Size,
        orientation: UIDeviceOrientation
    ) -> BoxRatio {
        let viewWidth = CGFloat(viewSize.width)
        let viewHeight = CGFloat(viewSize.height)

        // Normalize to viewSize
        let leftRatio = min(max(boxScaled.left / viewWidth, 0.0), 1.0)
        let topRatio = min(max(boxScaled.top / viewHeight, 0.0), 1.0)
        let widthRatio = min(max(boxScaled.width / viewWidth, 0.0), 1.0 - leftRatio)
        let heightRatio = min(max(boxScaled.height / viewHeight, 0.0), 1.0 - topRatio)

        switch orientation {
        case .portrait:
            return BoxRatio(leftRatio, topRatio, widthRatio, heightRatio)
        case .landscapeRight:
            return BoxRatio(
                topRatio,
                1 - leftRatio - widthRatio,
                heightRatio,
                widthRatio
            )
        case .portraitUpsideDown:
            return BoxRatio(
                1 - leftRatio - widthRatio,
                1 - topRatio - heightRatio,
                widthRatio,
                heightRatio
            )
        case .landscapeLeft:
            return BoxRatio(
                1 - topRatio - heightRatio,
                leftRatio,
                heightRatio,
                widthRatio
            )
        default:
            return BoxRatio(leftRatio, topRatio, widthRatio, heightRatio)
        }
    }

    static func scaleBoxBasedOnViewSize(box: BoundingBox, imageSize: Size, viewSize: Size) -> BoundingBox {
        let imageScaling = getImageScalingBasedOnViewSize(imageSize: imageSize, viewSize: viewSize)
        let (scale, dx, dy) = imageScaling.destructured()

        let left = CGFloat(box.left) * scale - dx
        let top = CGFloat(box.top) * scale - dy
        let right = CGFloat(box.right) * scale - dx
        let bottom = CGFloat(box.bottom) * scale - dy
        let width = right - left
        let height = bottom - top

        return BoundingBox(
            width: width,
            height: height,
            left: left,
            top: top,
            right: right,
            bottom: bottom
        )
    }

    static func filterBarcodes(
        barcodes: [Barcode],
        imageSize: Size,
        viewSize: Size,
        ratio: Ratio,
        imageRotation: UIImage.Orientation
    ) -> [Barcode] {
        let viewWidth = CGFloat(viewSize.width)
        let viewHeight = CGFloat(viewSize.height)

        let scanViewWidth = ratio.width * viewWidth
        let scanViewHeight = ratio.height * viewHeight

        // Calculate the current scan area for viewSize.
        let scanLeft = (viewWidth - scanViewWidth) / 2.0
        let scanTop = (viewHeight - scanViewHeight) / 2.0
        let scanRight = scanLeft + scanViewWidth
        let scanBottom = scanTop + scanViewHeight

        return barcodes.filter { barcode in
            let rect = self.getRectBasedOnRotation(rect: barcode.frame, imageSize: imageSize, imageRotation: imageRotation)
            let box = self.getBoundingBoxFromRect(rect: rect)
            let boxScaled = self.scaleBoxBasedOnViewSize(box: box, imageSize: imageSize, viewSize: viewSize)
            return boxScaled.left >= scanLeft &&
                   boxScaled.top >= scanTop &&
                   boxScaled.right <= scanRight &&
                   boxScaled.bottom <= scanBottom
        }
    }

    static func formatBarcode(
        barcode: Barcode,
        imageSize: Size,
        viewSize: Size,
        orientation: UIDeviceOrientation,
        imageRotation: UIImage.Orientation
    ) -> [String:Any] {
        var map : [String:Any] = [:]
        let rect = self.getRectBasedOnRotation(rect: barcode.frame, imageSize: imageSize, imageRotation: imageRotation)
        let box = self.getBoundingBoxFromRect(rect: rect)
        let boxScaled = self.scaleBoxBasedOnViewSize(box: box, imageSize: imageSize, viewSize: viewSize)
        let boxRatio = self.getBoxRatioBasedOnOrientationForUI(
            boxScaled: boxScaled,
            viewSize: viewSize,
            orientation: orientation
        )

        // Raw values
        map["width"] = boxScaled.width
        map["height"] = boxScaled.height
        map["left"] = boxScaled.left
        map["top"] = boxScaled.top
        map["right"] = boxScaled.right
        map["bottom"] = boxScaled.bottom

        // Normalized values
        map["leftRatio"] = boxRatio.leftRatio
        map["topRatio"] = boxRatio.topRatio
        map["widthRatio"] = boxRatio.widthRatio
        map["heightRatio"] = boxRatio.heightRatio

        let displayValue = barcode.displayValue
        map["displayValue"] = displayValue
        let rawValue = barcode.rawValue
        map["rawValue"] = rawValue

        let valueType = barcode.valueType
        switch valueType {
            case .wiFi:
                let ssid = barcode.wifi?.ssid
                map["ssid"] = ssid
                let password = barcode.wifi?.password
                map["password"] = password
                let encryptionType = barcode.wifi?.type
                map["encryptionType"] = encryptionType

            case .URL:
                let title = barcode.url!.title
                map["title"] = title
                let url = barcode.url!.url
                map["url"] = url

            default:
                break;
        }

        return map
    }
}

File: android/src/main/java/com/visioncamerabarcodesscanner/ScannerUtils.kt
package com.visioncamerabarcodesscanner

import android.graphics.Rect
import android.graphics.RectF

import com.facebook.react.bridge.ReadableNativeMap
import com.facebook.react.bridge.WritableNativeMap

import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ALL_FORMATS
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_AZTEC
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODABAR
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_128
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_39
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_93
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_DATA_MATRIX
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_13
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_8
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ITF
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_PDF417
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_QR_CODE
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UPC_A
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UPC_E

data class Size(
    val width: Int = 0,
    val height: Int = 0,
)

data class Ratio(
    val width: Float = 1f,
    val height: Float = 1f,
)

data class BoxRatio(
    val leftRatio: Double,
    val topRatio: Double,
    val widthRatio: Double,
    val heightRatio: Double,
)

data class Scaling(
    val scale: Double,
    val dx: Double,
    val dy: Double,
)

enum class Orientation (val value: String) {
    PORTRAIT("portrait"),
    LANDSCAPE_RIGHT("landscape-right"),
    PORTRAIT_UPSIDE_DOWN("portrait-upside-down"),
    LANDSCAPE_LEFT("landscape-left"),
}

object ScannerUtils {
    fun getOptionsRatio(options: MutableMap<String, Any>?): Ratio {
        val ratio = options?.get("ratio") as? Map<*, *>
        val width = ((ratio?.get("width") as? Number)?.toFloat() ?: 1f).coerceIn(0f, 1f)
        val height = ((ratio?.get("height") as? Number)?.toFloat() ?: 1f).coerceIn(0f, 1f)
        return Ratio(width, height)
    }

    fun getOptionsOrientation(options: MutableMap<String, Any>?): Orientation {
        return when (options?.get("orientation") as? String) {
            "portrait" -> Orientation.PORTRAIT
            "landscape-left" -> Orientation.LANDSCAPE_LEFT
            "portrait-upside-down" -> Orientation.PORTRAIT_UPSIDE_DOWN
            "landscape-right" -> Orientation.LANDSCAPE_RIGHT
            else -> Orientation.PORTRAIT
        }
    }

    fun getOptionsViewSize(options: MutableMap<String, Any>?): Size? {
        val viewSize = options?.get("viewSize") as? Map<*, *>
        val width = (viewSize?.get("width") as? Number)?.toInt()
        val height = (viewSize?.get("height") as? Number)?.toInt()
        return if (width != null && height != null) {
            Size(width, height)
        } else {
            null
        }
    }

    fun getOptionsBarcodeFormats(options: MutableMap<String, Any>?): List<*> {
        val formats = options?.get("formats") as? List<*>
        return if (formats.isNullOrEmpty()) {
            listOf("all")
        } else {
            formats
        }
    }

    fun getSafeBarcodeFormats(formats: List<*>): List<Int> {
        return formats.mapNotNull { format ->
            if (format is String) {
                when (format) {
                    "code_128" -> FORMAT_CODE_128
                    "code_39" -> FORMAT_CODE_39
                    "code_93" -> FORMAT_CODE_93
                    "codabar" -> FORMAT_CODABAR
                    "ean_13" -> FORMAT_EAN_13
                    "ean_8" -> FORMAT_EAN_8
                    "itf" -> FORMAT_ITF
                    "upc_e" -> FORMAT_UPC_E
                    "upc_a" -> FORMAT_UPC_A
                    "qr" -> FORMAT_QR_CODE
                    "pdf_417" -> FORMAT_PDF417
                    "aztec" -> FORMAT_AZTEC
                    "data_matrix" -> FORMAT_DATA_MATRIX
                    "all" -> FORMAT_ALL_FORMATS
                    else -> FORMAT_ALL_FORMATS
                }
            } else {
                FORMAT_ALL_FORMATS
            }
        }
    }

    // ViewSize is used for scaling in the code.
    // The UI side should always be assumed as portrait.
    fun getSafeViewSize(imageSize: Size, viewSize: Size?): Size {
        if (viewSize != null) {
            val width = minOf(viewSize.width, viewSize.height)
            val height = maxOf(viewSize.width, viewSize.height)
            return Size(width, height)
        } else {
            return imageSize
        }
    }

    fun getImageSizeBasedOnRotation(imageSizeRaw: Size, imageRotationDegrees: Int): Size {
        return when (imageRotationDegrees) {
            90, 270 -> Size(imageSizeRaw.height, imageSizeRaw.width)
            else -> imageSizeRaw
        }
    }

    private fun getImageScalingBasedOnViewSize(imageSize: Size, viewSize: Size): Scaling {
        val imageWidth = imageSize.width.toDouble()
        val imageHeight = imageSize.height.toDouble()
        val imageAspect = imageWidth / imageHeight

        val viewWidth = viewSize.width.toDouble()
        val viewHeight = viewSize.height.toDouble()
        val viewAspect = viewWidth / viewHeight

        val scale: Double
        val scaledImageWidth: Double
        val scaledImageHeight: Double
        val dx: Double
        val dy: Double
        if (imageAspect > viewAspect) {
            // Image is wider -> crop left/right
            scale = viewHeight / imageHeight
            scaledImageWidth = imageWidth * scale
            scaledImageHeight = viewHeight
            dx = (scaledImageWidth - viewWidth) / 2.0
            dy = 0.0
        } else if (viewAspect > imageAspect) {
            // Image is taller → crop top/bottom
            scale = viewWidth / imageWidth
            scaledImageWidth = viewWidth
            scaledImageHeight = imageHeight * scale
            dx = 0.0
            dy = (scaledImageHeight - viewHeight) / 2.0
        } else {
            // Full compatibility -> crop nothing
            // Default if user does not provide viewSize value
            scale = 1.0
            scaledImageWidth = viewWidth
            scaledImageHeight = viewHeight
            dx = 0.0
            dy = 0.0
        }

        return Scaling(scale, dx, dy)
    }

    private fun getBoxRatioBasedOnOrientationForUI(
        boxScaled: RectF,
        viewSize: Size,
        orientation: Orientation
    ): BoxRatio {
        val viewWidth = viewSize.width.toDouble()
        val viewHeight = viewSize.height.toDouble()

        // Normalize to viewSize
        val leftRatio = (boxScaled.left / viewWidth).coerceIn(0.0, 1.0)
        val topRatio = (boxScaled.top / viewHeight).coerceIn(0.0, 1.0)
        val widthRatio = (boxScaled.width() / viewWidth).coerceIn(0.0, 1.0 - leftRatio)
        val heightRatio = (boxScaled.height() / viewHeight).coerceIn(0.0, 1.0 - topRatio)

        return when (orientation) {
            Orientation.PORTRAIT -> BoxRatio(leftRatio, topRatio, widthRatio, heightRatio)
            Orientation.LANDSCAPE_RIGHT -> BoxRatio(
                topRatio,
                1 - leftRatio - widthRatio,
                heightRatio,
                widthRatio
            )
            Orientation.PORTRAIT_UPSIDE_DOWN -> BoxRatio(
                1 - leftRatio - widthRatio,
                1- topRatio - heightRatio,
                widthRatio,
                heightRatio
            )
            Orientation.LANDSCAPE_LEFT -> BoxRatio(
                1 - topRatio - heightRatio,
                leftRatio,
                heightRatio,
                widthRatio
            )
        }
    }

    private fun scaleBoxBasedOnViewSize(box: Rect, imageSize: Size, viewSize: Size): RectF {
        val (scale, dx, dy) = getImageScalingBasedOnViewSize(imageSize, viewSize)
        return RectF(
            (box.left * scale - dx).toFloat(),
            (box.top * scale - dy).toFloat(),
            (box.right * scale - dx).toFloat(),
            (box.bottom * scale - dy).toFloat()
        )
    }

    fun filterBarcodes(
        barcodes: List<Barcode>,
        imageSize: Size,
        viewSize: Size,
        ratio: Ratio
    ): List<Barcode> {
        val viewWidth = viewSize.width.toFloat()
        val viewHeight = viewSize.height.toFloat()

        val scanViewWidth = ratio.width * viewWidth
        val scanViewHeight = ratio.height * viewHeight

        // Calculate the current scan area for viewSize.
        val scanLeft = (viewWidth - scanViewWidth) / 2
        val scanTop = (viewHeight - scanViewHeight) / 2
        val scanRight = scanLeft + scanViewWidth
        val scanBottom = scanTop + scanViewHeight

        return barcodes.filter { barcode ->
            val box = barcode.boundingBox ?: return@filter false
            val boxScaled = scaleBoxBasedOnViewSize(box, imageSize, viewSize)
            return@filter boxScaled.left >= scanLeft &&
                    boxScaled.top >= scanTop &&
                    boxScaled.right <= scanRight &&
                    boxScaled.bottom <= scanBottom
        }
    }

    fun formatBarcode(
        barcode: Barcode,
        imageSize: Size,
        viewSize: Size,
        orientation: Orientation
    ): ReadableNativeMap {
        val map = WritableNativeMap()
        val box = barcode.boundingBox

        if (box != null) {
            val boxScaled = scaleBoxBasedOnViewSize(box, imageSize, viewSize)
            val boxScaledRatio = getBoxRatioBasedOnOrientationForUI(boxScaled, viewSize, orientation)

            // Raw values
            map.putInt("width", boxScaled.width().toInt())
            map.putInt("height", boxScaled.height().toInt())
            map.putInt("left", boxScaled.left.toInt())
            map.putInt("top", boxScaled.top.toInt())
            map.putInt("right", boxScaled.right.toInt())
            map.putInt("bottom", boxScaled.bottom.toInt())

            // Normalized values
            map.putDouble("leftRatio", boxScaledRatio.leftRatio)
            map.putDouble("topRatio", boxScaledRatio.topRatio)
            map.putDouble("widthRatio", boxScaledRatio.widthRatio)
            map.putDouble("heightRatio", boxScaledRatio.heightRatio)
        }

        val rawValue = barcode.rawValue
        map.putString("rawValue", rawValue)
        val displayValue = barcode.displayValue
        map.putString("displayValue", displayValue)
        val valueType = barcode.valueType

        when (valueType) {
            Barcode.TYPE_WIFI -> {
                val ssid = barcode.wifi!!.ssid
                map.putString("ssid", ssid)
                val password = barcode.wifi!!.password
                map.putString("password", password)
                val encryptionType = barcode.wifi!!.encryptionType
                map.putInt("encryptionType", encryptionType)
            }

            Barcode.TYPE_URL -> {
                val title = barcode.url!!.title
                map.putString("title", title)
                val url = barcode.url!!.url
                map.putString("url", url)
            }
        }

        return map
    }
}
